# 01 - Data Audit
## Product Placement Optimization - Nepali Grocery Store

### Purpose
This notebook performs a complete audit of the raw transaction dataset 
obtained from a Nepali grocery retail store. The goal is to understand 
the dataset structure, confirm data completeness, and identify quality 
issues before any cleaning or analysis begins.

### Dataset
- **Source:** Real POS transaction data from a Nepali grocery retail store
- **Size:** 768,222 rows x 14 columns
- **Period:** July 2025 to May 2026 (10 months)

In [ ]:
# Importing necessary libraries
import pandas as pd

In [ ]:
#Loading raw data
df = pd.read_excel(r'C:\Users\swara\Documents\data enginnering projects\ProductPlacement\data\raw\sales_data_raw.xlsx', header=7)
print(f"Data loaded!, Structure: {len(df)} rows and {len(df.columns)} columns.")

Data loaded!, Structure: 768222 rows and 14 columns.


In [37]:
#First look
print(f"Columns: {list(df.columns)}")
print(f"\nFirst 5 rows:")
df.head()

Columns: ['Date', 'Miti', 'Inv.No', 'Customer', 'Product Group', 'Product', 'Uom', 'Qty', 'Rate', 'B. Amount', 'DISCOUNT', 'VAT', 'ROUND OFF', 'Total Amount']

First 5 rows:


,Date,Miti,Inv.No,Customer,Product Group,Product,Uom,Qty,Rate,B. Amount,DISCOUNT,VAT,ROUND OFF,Total Amount
0,2025-07-17,01/04/2082,ABT-000020,CASH A/C,POOJA ITEMS,Til Oil 500ml,PCS,1.0,175.0,154.867256,0.0,20.132744,0.0,175.0
1,2025-07-17,01/04/2082,ABT-000029,CASH A/C,GAULE,Gaule Chukamilo,PCS,1.0,105.0,105.000000,0.0,0.000000,0.0,105.0
2,2025-07-17,01/04/2082,ABT-000029,CASH A/C,ESSENTIAL FOOD ITEM,Krishma Mixnut 100gm,PCS,1.0,80.0,70.796460,0.0,9.203540,0.0,80.0
3,2025-07-17,01/04/2082,ABT-000029,CASH A/C,ESSENTIAL FOOD ITEM,Mong Khosta,KG,4.0,185.0,740.000000,0.0,0.000000,0.0,740.0
4,2025-07-17,01/04/2082,ABT-000029,CASH A/C,ESSENTIAL FOOD ITEM,Sugar,KG,5.0,105.0,464.601769,0.0,60.398231,0.0,525.0


In [40]:
# Key numbers exploration
print("KEY NUMBERS ")
print("-" * 50)
print(f"Total unique invoices: {df['Inv.No'].nunique():,}")
print(f"Total unique products: {df['Product'].nunique():,}")
print(f"Unique product groups: {df['Product Group'].nunique():,}")
print(f"Unique customers:      {df['Customer'].nunique():,}")
print(f"Date range:            {df['Date'].min()} to {df['Date'].max()}")

KEY NUMBERS 
--------------------------------------------------
Total unique invoices: 218,037
Total unique products: 5,681
Unique product groups: 38
Unique customers:      25
Date range:            2025-07-17 00:00:00 to 2026-05-20 00:00:00


In [110]:
# Product groups analysis
print("ALL PRODUCT GROUPS:")
print("-" * 70)
for i, (group,count) in enumerate(df["Product Group"].value_counts().items(),1):
    pct = (count / len(df)) * 100
    print(f"{i:2d} {group:40s} | {count:10} rows | {pct:5.1f}% |")

ALL PRODUCT GROUPS:
----------------------------------------------------------------------
 1 ESSENTIAL FOOD ITEM                      |     170669 rows |  22.2% |
 2 CANNED AND PACK GOODS                    |      96889 rows |  12.6% |
 3 SOAP AND CLEANER                         |      50868 rows |   6.6% |
 4 BISCUITS                                 |      47062 rows |   6.1% |
 5 CHOCOLATE & CANDY                        |      44774 rows |   5.8% |
 6 TEA & MASALA ITEM                        |      44202 rows |   5.8% |
 7 Cooking Oil                              |      37265 rows |   4.9% |
 8 HOUSEHOLD ITEM                           |      36091 rows |   4.7% |
 9 NOODLES ITEMS                            |      31964 rows |   4.2% |
10 SOFT DRINK                               |      30832 rows |   4.0% |
11 PERSONAL CARE                            |      29669 rows |   3.9% |
12 DAIRY PRODUCT                            |      29506 rows |   3.8% |
13 Rice                          

In [ ]:
# Top 30 products
print("TOP 30 MOST SOLD PRODUCTS:")
print("-" * 70)
for i, (products,count) in enumerate(df['Product'].value_counts().head(30).items(),1):
    print(f"{i:2d}. {products:40s} | sold{count:>7,} times |")

TOP 30 MOST SOLD PRODUCTS:
----------------------------------------------------------------------
 1. Sugar                                    | sold 23,106 times |
 2. Rato Dal                                 | sold 12,857 times |
 3. SALT                                     | sold  7,756 times |
 4. Chana                                    | sold  7,706 times |
 5. Local Milk 500ml                         | sold  7,222 times |
 6. COKE FANTA SPRITE 1.5L                   | sold  6,726 times |
 7. Kalo Dal                                 | sold  6,290 times |
 8. WAI WAI  CHICKEN                         | sold  6,171 times |
 9. Khukuri FT                               | sold  5,632 times |
10. RARA                                     | sold  5,341 times |
11. LUV POUCH DAHI 500ML                     | sold  5,209 times |
12. PATANJALI SOAP                           | sold  4,764 times |
13. Makai                                    | sold  4,760 times |
14. BADAM RAMRO                

In [121]:
# Missing values check
print("\n1. MISSING VALUES PER COLUMN:")
nulls = df.isnull().sum()
for col, n in nulls.items():
    status = "PROBLEM" if n>0 else "OK"
    print(f"{col:25s} {n:3d} missing | {status}")


1. MISSING VALUES PER COLUMN:
Date                        1 missing | PROBLEM
Miti                        1 missing | PROBLEM
Inv.No                      1 missing | PROBLEM
Customer                    0 missing | OK
Product Group               1 missing | PROBLEM
Product                     1 missing | PROBLEM
Uom                         1 missing | PROBLEM
Qty                         1 missing | PROBLEM
Rate                        1 missing | PROBLEM
B. Amount                   1 missing | PROBLEM
DISCOUNT                  158 missing | PROBLEM
VAT                       158 missing | PROBLEM
ROUND OFF                 158 missing | PROBLEM
Total Amount                1 missing | PROBLEM


In [122]:
# Duplicates check
dupes = df.duplicated().sum()
print(f"\n2. DUPLICATE ROWS: {dupes:,}")


2. DUPLICATE ROWS: 1,040


In [123]:
# Negative values
neg_qty = (df['Qty'] <= 0).sum()
neg_amount = (df['Total Amount'] <= 0).sum()
print(f"\n3. ROWS WITH QTY <= 0:          {neg_qty:,}")
print(f"   ROWS WITH TOTAL AMOUNT <= 0: {neg_amount:,}")


3. ROWS WITH QTY <= 0:          0
   ROWS WITH TOTAL AMOUNT <= 0: 0


In [126]:
# NaN in product group
nan_pg = df['Product Group'].isna().sum()
print(f"\n4. ROWS WITH NO PRODUCT GROUP:  {nan_pg:,}")



4. ROWS WITH NO PRODUCT GROUP:  1


# Audit Complete

### Summary of Findings
| Issue | Count | Action Required |
|-------|-------|----------------|
| Duplicate rows | 1,040 | Remove in cleaning |
| Missing product group | 1 | Remove in cleaning |
| Blank row | 1 | Remove in cleaning |
| Missing DISCOUNT/VAT/ROUND OFF | 158 | Investigate in cleaning |
| Inconsistent category names | Multiple | Standardize in cleaning |

### Next Step
Proceed to **02_data_cleaning.ipynb** to fix all issues identified above.